evaluator.go

```diff
@@ -0,0 +1,278 @@
+				splitValue := n.Value
+				if val == splitValue {
+					current = n.LeftChild
+				} else {
+					current = n.LeftChild + 1
+				}
+			}
+		}
+	}
+	return preds
+}
+
+type PiecewiseFlatForest struct {
+	Trees []*FlatTree `json:"trees"`
+}
+
+func NewPiecewiseFlatForest(forest *Forest) *PiecewiseFlatForest {
+	p := &PiecewiseFlatForest{make([]*FlatTree, len(forest.Trees))}
+	for i, n := range forest.Trees {
+		p.Trees[i] = NewFlatTree(n.Root)
+	}
+	return p
+}
+
+func (p *PiecewiseFlatForest) EvaluateNum(fm *FeatureMatrix) []float64 {
+	sz := fm.Data[0].Length()
+	n := float64(len(p.Trees))
+	preds := make([]float64, sz)
+	for _, tree := range p.Trees {
+		for i, pred := range tree.EvaluateNum(fm) {
+			preds[i] += pred / n
+		}
+	}
+	return preds
+}
+
+func (p *PiecewiseFlatForest) EvaluateCat(fm *FeatureMatrix) []string {
+	sz := fm.Data[0].Length()
+	bb := NewCatBallotBox(sz)
+	for _, tree := range p.Trees {
+		for i, pred := range tree.EvaluateCat(fm) {
+			bb.Vote(i, pred, 1.0)
+		}
+	}
+
+	preds := make([]string, sz)
+	for i := 0; i < sz; i++ {
+		preds[i] = bb.Tally(i)
+	}
+	return preds
+}
+
+type ContiguousFlatForest struct {
+	Roots []uint32
+	Nodes []*FlatNode
+}
+
+func NewContiguousFlatForest(forest *Forest) *ContiguousFlatForest {
+	var roots []uint32
+	var nodes []*FlatNode
+	for _, tree := range forest.Trees {
+		idx := uint32(len(nodes))
+		roots = append(roots, idx)
+		for _, node := range NewFlatTree(tree.Root).Nodes {
+			node.LeftChild += idx
+			nodes = append(nodes, node)
+		}
+	}
+	return &ContiguousFlatForest{
+		Roots: roots,
+		Nodes: nodes,
+	}
+}
+
+func (c *ContiguousFlatForest) EvaluateNum(fm *FeatureMatrix) []float64 {
+	sz := fm.Data[0].Length()
+	preds := make([]float64, sz)
+	for i := 0; i < sz; i++ {
+		result := 0.0
+		for _, root := range c.Roots {
+			current := root
+			for {
+				n := c.Nodes[current]
+				if n.Feature == leafFeature {
+					// im a leaf
+					val, _ := strconv.ParseFloat(n.Value, 64)
+					result += val
+					break
+				}
+				switch f := fm.Data[n.Feature].(type) {
+				case *DenseNumFeature:
+					val := f.NumData[0]
+					splitValue, _ := strconv.ParseFloat(n.Value, 64)
+					if val < splitValue {
+						current = n.LeftChild
+					} else {
+						current = n.LeftChild + 1
+					}
+				case *DenseCatFeature:
+					val := f.GetStr(0)
+					splitValue := n.Value
+					if val == splitValue {
+						current = n.LeftChild
+					} else {
+						current = n.LeftChild + 1
+					}
+				}
+			}
+		}
+		preds[i] = result / float64(len(c.Roots))
+	}
+	return preds
+}
+
+func (c *ContiguousFlatForest) EvaluateCat(fm *FeatureMatrix) []string {
+	sz := fm.Data[0].Length()
+	bb := NewCatBallotBox(sz)
+
+	for i := 0; i < sz; i++ {
+		for _, root := range c.Roots {
+			current := root
+			for {
+				n := c.Nodes[current]
+				if n.Feature == leafFeature {
+					// im a leaf
+					bb.Vote(i, n.Value, 1.0)
+					break
+				}
+				switch f := fm.Data[n.Feature].(type) {
+				case *DenseNumFeature:
```

Comment:
This switch seems like it's a slower operation to person in the middle of a loop, and wondering why you check for DenseNum in evaluating Cats. Does your tree fix cats/nums? Looks like it's possible, but not in use?